### DISCUSSION TITLE
Ideas for an improved branch naming scheme

### DISCUSSION THREAD

**Discussion by Gornoka** [united-manufacturing-hub/united-manufacturing-hub#800] [Ideas]
# Hello everyone.

I think it may be time to improve the branch naming scheme that we are currently using in the umh repo.

Currently the following scheme is used in the repo.
````
<isssue_id>-<classic_branch_name>
````

I want to propose the following scheme, which combines the ideas of conventional commits with the current scheme of issue number based branches. 
Every branch upon creation should already define a goal for the branch, 
and not just for the individual commits. 
obviously not every commit inside of a branch will match the same commit type but this should not be a problem.

## Branch names

Branch names follow a similar convention as commits.
````
<type>/<issue_id>/<normal_name>
````
### Types

The following types are allowed for branches.
Nested Types are allowed.

* Feature  
* Fix
* Docs
* Chore
* Action
* Merge

#### Examples
````
Feature/123/dynamic_rendering_of_images
Chore/345/css_rework
Fix/678/migrations_are_broken_again111
Merge/Quick/fix_random_merge_conflict
````

### Issue id
Issue id is supposed to be the issue number.

For single commit branches it is fine to use the quick keyword
these should only be used for minor text fixes or stuff like that.

examples:
````
Feature/123/Hello_issue
Docs/Quick/minor_text_fixes
````

### Normal name
This is what you would name your branch if you were working alone, e.g. the thing that you wanted to do / actually did.

Do not use reserved names `master` `staging` `main` .

Do not use anything that could make problems with external tooling ( special charchters , `*.+*//üäöø` etc.
Everything else is ok, but please use common sense.
(No ascii art or stuff like that)

#### examples

````
# ok
Feature/123/Hello_issue
Docs/Quick/minor_text_fixes
# not ok
Feature/123/python_changessssss
Feature/123/.*.*.*.*.new_snowflake_design_for_christmas
````

## Reasoning

* It is immediately obvious from the branch name what a branch is supposed to accomplish
* it is immediately obvious which issue a branch is associated with (also in old system)
* it is easy to ditinguish which part of the issue is the improvement type and which part is the number and name
* it makes sorting for specific things easier since you are now able to search for something by intention and not just by name
* External tooling supports this as a folder structure to make life easier when searching and reviewing code
* full backwards compatibility, it does not interfere with the primary branches or tags
* eventually the future pipelines can incorporate this additional context (e.g. special tests etc.)

## Points that most likely will need discussion and adaptation for the umh
* Issue types and issue type definition 
* normal name guidelines
* implications on existing pipelines ( e.g. labeling, testing and linting)
* allowed nesting levels ( is it usefull for your workflow to have this additional layer of verbosity ? )
  - e.g. for something like the action and merge type it may make more sense to split them up further for the Feature type maybe less

# 

I am interested in your feedback / ideas.

With kind regards 
Lukas Hennies

**Comment by Scarjit** [comment]
Tool integration inside GoLand
![goland64_SpH2SlUsXw](https://user-images.githubusercontent.com/6317420/149959835-5e73191c-6533-4dfb-ab8b-a0406e5c432c.png)
 
Folders will only show, inside the "Git" tab, not inside the quick change tab
